File: wimg-ios/wimgWidget/WidgetData.swift
import Foundation

struct WidgetData {
    let available: Double
    let income: Double
    let expenses: Double
    let savingsRate: Int
    let nextMerchant: String?
    let nextAmount: Double?
    let nextDate: String?
    let recentTransactions: [(desc: String, amount: Double, date: String)]
    let updatedAt: Date
    /// User opted to hide € amounts in widgets. Stored in App Group defaults
    /// at `wimg_widget_mask_amounts` and read both here and by the views.
    let maskAmounts: Bool

    static let empty = WidgetData(
        available: 0, income: 0, expenses: 0, savingsRate: 0,
        nextMerchant: nil, nextAmount: nil, nextDate: nil,
        recentTransactions: [],
        updatedAt: .distantPast,
        maskAmounts: false
    )

    static func load() -> WidgetData {
        let defaults = UserDefaults(suiteName: "group.com.wimg.app")
        let mask = defaults?.bool(forKey: "wimg_widget_mask_amounts") ?? false
        guard let defaults,
              let jsonString = defaults.string(forKey: "wimg_widget_data"),
              let data = jsonString.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return WidgetData(
            available: 0, income: 0, expenses: 0, savingsRate: 0,
            nextMerchant: nil, nextAmount: nil, nextDate: nil,
            recentTransactions: [], updatedAt: .distantPast,
            maskAmounts: mask
        ) }

        return WidgetData(
            available: dict["available"] as? Double ?? 0,
            income: dict["income"] as? Double ?? 0,
            expenses: dict["expenses"] as? Double ?? 0,
            savingsRate: dict["savings_rate"] as? Int ?? 0,
            nextMerchant: dict["next_recurring_merchant"] as? String,
            nextAmount: dict["next_recurring_amount"] as? Double,
            nextDate: dict["next_recurring_date"] as? String,
            recentTransactions: {
                guard let txArray = dict["recent_transactions"] as? [[String: Any]] else { return [] }
                return txArray.prefix(5).map { tx in
                    (desc: tx["description"] as? String ?? "",
                     amount: tx["amount"] as? Double ?? 0,
                     date: tx["date"] as? String ?? "")
                }
            }(),
            updatedAt: Date(timeIntervalSince1970: dict["updated_at"] as? Double ?? 0),
            maskAmounts: mask
        )
    }

    var hasData: Bool { updatedAt != .distantPast }

    var nextDateFormatted: String? {
        guard let nextDate else { return nil }
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        guard let date = formatter.date(from: nextDate) else { return nil }
        let lang = UserDefaults(suiteName: "group.com.wimg.app")?.string(forKey: "wimg_locale")
            ?? UserDefaults.standard.string(forKey: "wimg_locale")
            ?? "de"
        let display = DateFormatter()
        display.dateFormat = "d. MMM"
        display.locale = Locale(identifier: lang == "en" ? "en_US" : "de_DE")
        return display.string(from: date)
    }
}

File: wimg-android/app/src/main/java/com/wimg/app/widget/WidgetData.kt
package com.wimg.app.widget

import android.content.Context
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

data class RecentTx(val description: String, val amount: Double)

data class WidgetData(
    val available: Double = 0.0,
    val income: Double = 0.0,
    val expenses: Double = 0.0,
    val savingsRate: Int = 0,
    val nextMerchant: String? = null,
    val nextAmount: Double? = null,
    val nextDate: String? = null,
    val recent: List<RecentTx> = emptyList(),
    val hasData: Boolean = false,
) {
    companion object {
        fun load(context: Context): WidgetData {
            val prefs = context.getSharedPreferences("wimg_widget", Context.MODE_PRIVATE)
            val jsonStr = prefs.getString("wimg_widget_data", null) ?: return WidgetData()
            return try {
                val obj = Json.parseToJsonElement(jsonStr).jsonObject
                WidgetData(
                    available = obj["available"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                    income = obj["income"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                    expenses = obj["expenses"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                    savingsRate = obj["savings_rate"]?.jsonPrimitive?.intOrNull ?: 0,
                    nextMerchant = obj["next_recurring_merchant"]?.jsonPrimitive?.contentOrNull,
                    nextAmount = obj["next_recurring_amount"]?.jsonPrimitive?.doubleOrNull,
                    nextDate = obj["next_recurring_date"]?.jsonPrimitive?.contentOrNull,
                    recent = obj["recent"]?.jsonArray?.mapNotNull { el ->
                        val o = el.jsonObject
                        val d = o["desc"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
                        val a = o["amount"]?.jsonPrimitive?.doubleOrNull ?: return@mapNotNull null
                        RecentTx(d, a)
                    } ?: emptyList(),
                    hasData = true,
                )
            } catch (_: Exception) {
                WidgetData()
            }
        }
    }
}
